src/node/hooks/express/openapi.js

```diff
@@ -0,0 +1,671 @@
+
+    for (const style of [APIPathStyle.FLAT, APIPathStyle.REST]) {
+      const apiRoot = getApiRootForVersion(version, style);
+
+      // generate openapi definition for this API version
+      const definition = generateDefinitionForVersion(version, style);
+
+      // serve version specific openapi definition
+      app.get(`${apiRoot}/openapi.json`, (req, res) => {
+        // For openapi definitions, wide CORS is probably fine
+        res.header('Access-Control-Allow-Origin', '*');
+        res.json({ ...definition, servers: [generateServerForApiVersion(apiRoot, req)] });
+      });
+
+      // serve latest openapi definition file under /api/openapi.json
+      const isLatestAPIVersion = version === apiHandler.latestApiVersion;
+      if (isLatestAPIVersion) {
+        app.get(`/${style}/openapi.json`, (req, res) => {
+          res.header('Access-Control-Allow-Origin', '*');
+          res.json({ ...definition, servers: [generateServerForApiVersion(apiRoot, req)] });
+        });
+      }
+
+      // build openapi-backend instance for this api version
+      const api = new OpenAPIBackend({
+        apiRoot, // each api version has its own root
+        definition,
+        validate: false,
+        // for a small optimisation, we can run the quick startup for older
+        // API versions since they are subsets of the latest api definition
+        quick: !isLatestAPIVersion,
+      });
+
+      // register default handlers
+      api.register({
+        notFound: (c, req, res) => {
+          res.statusCode = 404;
+          res.send({ code: 3, message: 'no such function', data: null });
+        },
+        notImplemented: (c, req, res) => {
+          res.statusCode = 501;
+          res.send({ code: 3, message: 'not implemented', data: null });
+        },
+      });
+
+      // register operation handlers (calls apiHandler.handle)
+      for (const funcName in apiHandler.version[version]) {
+        const handler = async (c, req, res) => {
+          // parse fields from request
+          const { header, params, query } = c.request;
+
+          // read form data if method was POST
+          let formData = {};
+          if (c.request.method === 'post') {
+            const form = new formidable.IncomingForm();
+            const parseForm = promisify(form.parse).bind(form);
+            formData = await parseForm(req);
+          }
+
+          const fields = Object.assign({}, header, params, query, formData);
+
+          // log request
+          apiLogger.info(`REQUEST, v${version}:${funcName}, ${JSON.stringify(fields)}`);
+
+          // pass to api handler
+          let data = await apiHandler.handle(version, funcName, fields, req, res);
+          if (!data) {
+            data = null;
+          }
+
+          // return in common format
+          let response = { code: 0, message: 'ok', data };
+
+          // log response
+          apiLogger.info(`RESPONSE, ${funcName}, ${JSON.stringify(response)}`);
```

Comment:
A doubt: what does happen when there is a failure? Is the API going to return a non zero code (and a relevant message)?
I am not sure the tests cover this case, and from the look of this part there could be this risk.

For your reference, the documentation for the HTTP API response format is in [doc/api/http_api.md](https://github.com/ether/etherpad-lite/blob/4ee5ddb496bdd3ab107efa26a001a87ac343fd76/doc/api/http_api.md#response-format), this is the relevant section:

> ### Response Format
> Responses are valid JSON in the following format:
> 
> ```json
> {
>   "code": number,
>   "message": string,
>   "data": obj
> }
> ```
> 
> * **code** a return code
>   * **0** everything ok
>   * **1** wrong parameters
>   * **2** internal error
>   * **3** no such function
>  * **4** no or wrong API Key
> * **message** a status message. It's ok if everything is fine, else it contains an error message
> * **data** the payload

Could you please confirm if we are respecting this?

Reply:
Yes, we cover all the same cases as the previous implementation. 

1. If the api function throws an error, we either pass the custom error thrown by the function as code 1 or code 2 for non-`apierror` exceptions.

Previously: https://github.com/ether/etherpad-lite/blob/master/src/node/handler/APIHandler.js#L209-L219
Now: https://github.com/ether/etherpad-lite/blob/427d388631fceeacdec4dbc97933f35540655bf2/src/node/hooks/express/openapi.js#L650-L659

2. If the api definition doesn't contain the route being called, we return code 3 for no such function using the `notFound` handler:

https://github.com/ether/etherpad-lite/blob/427d388631fceeacdec4dbc97933f35540655bf2/src/node/hooks/express/openapi.js#L585-L588

3. For code 4, `Unauthorized`, there's a test which we pass

Test: https://github.com/ether/etherpad-lite/blob/master/tests/backend/specs/api/pad.js#L51-L59

Reply:
ok, thanks